### PR TITLE
Blocks: Shortcode: Support multi-line shortcodes

### LINKED ISSUE #4456: New lines in shortcode content are ignored
## Issue Overview

The new editor does not respect new lines in shortcode content. It's impossible to use the new editor to edit posts which depend on this. For me this is a blocker; it breaks my sites.

I have a number of shortcodes where each line in the content is expected to be processed separately. With Gutenberg v2.0.0 it's impossible to safely edit a post containing these shortcodes without the content being broken.  Take for example the [bw_csv] shortcode which converts CSV content into a table.

```
[bw_csv]One,Two,Three
A,B,C
D,E,F
[/bw_csv]
```
During shortcode expansion the first line becomes table the table heading.
Subsequent lines are table rows. So the expected output is like this.

One | Two | Three
--- | ---- | ----
A | B | C
D | E | F

When the new lines are removed by the editor the shortcode does not expand correctly. Everything is treated as the table heading. 

The basic problem can be demonstrated even without the shortcode being active.
In the current editor you can quite happily switch between Visual and Text and the site is not broken.  
Once the content is converted the new lines in the original source are lost. 

## Steps to Reproduce (for bugs)

1. Create a new post using Posts > All Posts > Add New > Classic Editor
2. Enter content as in the shortcode above
3. View the post.
4. The content should appear as typed.

Repeat the above steps for another new post
5. Now Edit the post using Gutenberg
6. The new line characters do not survive the conversion.
7. Any attempt to recreate the original content using Gutenberg will fail.
8. Switching to the Code editor to remove unwanted paragraphs ( #3900 ) doesn't work as that ignores new lines as well. They disappear as soon as you click on Update.

I've also tried using different block types: Shortcode, Custom HTML. 

## Expected Behavior

- The editor should not be removing new lines.
- It should not add anything either.
- It works in WordPress 4.9.1
- In my opinion the Shortcode block should support content created AS IS. It should respect any new line characters in the content. 

## Current Behavior
I've been unable to find any existing block that respects my new lines. 
- Custom HTML strips the new lines on save.
- The shortcode block also strips new lines.

## Possible Solution
Shortcode block should respect the content as typed.

## Screenshots / Video

image
image
image

## Related Issues and/or PRs

#3900 
#1709 - see https://github.com/WordPress/gutenberg/issues/1709#issuecomment-321574711

## Todos
- [ ] Tests
- [ ] Documentation

### PR DESCRIPTION
**Pull request by mcsf**
## Description
Partly fixes #4456 

~_In progress: this doesn't yet preserve newlines within a Shortcode block if reloading the block (when switching between Visual and Code modes or by reloading Gutenberg)._~

When converting content to blocks (either explicitly via the _Convert to blocks_ action or by pasting content), content fragments that are recognized as shortcodes (but don't have a specific block to handle it) are then sent through the generic Shortcode block's `'shortcode'` transform. Since raw content is canonically treated as HTML, the Shortcode transform receives any newline as `<br />`. This PR converts these into `\n` inside the shortcode transform.

This was an issue because the Shortcode block _intentionally_ treats its contents literally. This is because it is meant to be a place where HTML entities are not decoded, so as to better support the extent of WordPress shortcode usage. See #3609 for history.

## How Has This Been Tested?
1. Attempt to repeat the reported bug by following the steps in #4456. In a nutshell, try to _convert from classic to blocks_ or try to _paste_ content with a multi-line shortcode such as:

```
[bwcsv]One,Two,Three
A,B,C
D,E,F
[/bwcsv]
```

and make sure the process creates a Shortcode block that respects the original line breaks.

Caveat: this is a remaining issue to be solved concerning pasting for shortcodes containing underscores. In other words, the issue author's `[bw_csv]` shortcode will not be correctly parsed upon pasting. It will, however, be handled correctly when converting from classic.

2. Make sure that one-line shortcodes are still supported.
3. Make sure that a fragment of *text* that reads `<br />` will not be treated as a newline.

## Screenshots (jpeg or gifs if applicable):

Before | After
--|--
Before | After

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style.
- [ ] My code has proper inline documentation.

### CODE REVIEW

**Comment by mcsf** [comment]
The main hurdle here is that, during content serialization, line breaks inside the shortcode block are destroyed by the _beautifier_:

https://github.com/WordPress/gutenberg/blob/62d746f98d73b3c1a6ccbcc39ee9388d5a4d8829/blocks/api/serializer.js#L199

Now, there's ways to handle this. One is that there may be other block types out there where total control of a block's inner HTML is desired. For this, something like `supports` would (and does, on my test site) work:

```js
// in core-blocks/shortcode/index.js
supports: { …, rawContent: true }

// in `blocks/api/serializer.js`'s `getBlockContent`:
return (
	getUnknownTypeHandlerName() === block.name ||
	hasBlockSupport( blockType, 'rawContent' ) ||
	! saveContent
) ? saveContent : getBeautifulContent( saveContent );
```

However, with @aduth and @iseulde we've wondered whether it still made sense, going forward, to systematically beautify blocks' inner HTML. If it doesn't, then we shouldn't rush to add a new `support` term. Should we then add an exception for shortcode, e.g.

```diff
+++ core-blocks/index.js
 setDefaultBlockName( paragraph.name );
 setUnknownTypeHandlerName( freeform.name );
+addRawContentBlockType( shortcode.name );
```

```js
return (
	getUnknownTypeHandlerName() === block.name ||
	isRawContentBlockType( block.name ) ||
  	! saveContent
) ? saveContent : getBeautifulContent( saveContent );
```

?

**Comment by mcsf** [comment]
The last two commits work and represent the two approaches I described above.

**Comment by aduth** [comment]
In general I wish we'd just solve this from the serializer (no additional processing on `RawHTML`) and drop the beautifier altogether. I remarked in https://github.com/WordPress/gutenberg/issues/4456#issuecomment-381592662 that I have this mostly working, but some questions remain on exactly what format the indentation should take.

Which is to say... can you tell me what beauty means to you? :troll:

**Comment by mcsf** [comment]
> Which is to say... can you tell me what beauty means to you? :troll:

I think we all know that the Beautiful is found only in LISP-style S-expressions.

More seriously:

> In general I wish we'd just solve this from the serializer (no additional processing on `RawHTML`)

This makes sense, and what the comment you pointed to suggests seems reasonable to me. What action items do we have? How soon can we work on the serializer?

**Comment by aduth** [comment]
There is actually a `--preserve-newlines` option (which I assume has a JS property equivalent), which may be fine as a short-term resolution:

https://github.com/beautify-web/js-beautify#css--html

To be added as an option to `getBeautifulContent`:

https://github.com/WordPress/gutenberg/blob/f037e9028fff2a7ff1353db8f113ccfa358ded5b/blocks/api/serializer.js#L168-L174

Since the element serializer outputs no newlines of its own (except for those hard-coded), I expect this should work without conflict.

File: core-blocks/shortcode/index.js [hunk 1]
```diff
@@ -45,7 +44,7 @@ export const settings = {
 					text: {
 						type: 'string',
 						shortcode: ( attrs, { content } ) => {
-							return content;
+							return content.replace( /<br \/>/g, '' );
```
Comment (iseulde) [inline core-blocks/shortcode/index.js]: How do we know if these are added by the browser? Guess we would need to check if it's plain text, and only remove then?

Btw, the current editor also destroys the following.

```html
[bwcsv]One,Two,Three
A,B,C
D,E,F
[/bwcsv]
```

becomes

```html
<p>[bwcsv]One,Two,Three A,B,C D,E,F [/bwcsv]</p>
```

Comment (iseulde) [reply core-blocks/shortcode/index.js]: And if pasted in TinyMCE it becomes:

```html
<p>[bw_csv]One,Two,Three<br />A,B,C<br />D,E,F<br />[/bw_csv]</p>
```

Comment (mcsf) [reply core-blocks/shortcode/index.js]: > Btw, the current editor also destroys the following.

Not if manually entered, right?

Regardless, a block whose purpose is to hold shortcodes should be capable of preserving arbitrary formatting, IMO.

Comment (iseulde) [reply core-blocks/shortcode/index.js]: Yes, it should preserve it, agreed. But this line here is about input. Now I just realised this is actually the Markdown converter inserting these line break elements? With the `simpleLineBreaks` option. 😰

**Comment by aduth** [comment]
`preserve_newlines` is [defaulted to `true`](https://github.com/beautify-web/js-beautify/blob/c041b9c775df3692852b25df281c4e198c591817/js/lib/beautify-html.js#L68) for HTML beautification, so this unfortunately doesn't seem to be helping the original issue.

**Comment by aduth** [comment]
Note: Alternative solution at #6716.

**Comment by mcsf** [comment]
I'm happier with a broader fix like #6716, so I'll close this one. If we change our minds, we can open it back.

**Pull request closed without merging**